chain/chain/src/store.rs

```diff
@@ -1979,7 +1979,7 @@ impl<'a> ChainStoreUpdate<'a> {
         self.tail = Some(height);
         let genesis_height = self.get_genesis_height();
         // When fork tail is behind tail, it doesn't hurt to set it to tail for consistency.
-        if self.fork_tail.unwrap_or(genesis_height) < height {
+        if self.fork_tail().unwrap_or(genesis_height) < height {
```

Comment:
this is the culprit of the problem .. ;-)

All the other changes here are test improvements.

Reply:
Hm, now that this is a Result rather than an option, I wonder if this should use `?` rather than `.unwrap_or_else`, to not mask db failures and such.

The `or_else` already happens in the implementation of the method:

```
    fn fork_tail(&self) -> Result<BlockHeight, Error> {
        self.store
            .get_ser(ColBlockMisc, FORK_TAIL_KEY)
            .map(|option| option.unwrap_or_else(|| self.genesis_height))
            .map_err(|e| e.into())
    }
```

So I think error here always means catastrophic erro